src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -274,40 +293,100 @@ internal GeneratorDriverState RunGeneratorsCore(Compilation compilation, Diagnos
+            var driverStateBuilder = new DriverStateTable.Builder(_state.StateTable, cancellationToken);
+            driverStateBuilder.SetInputState(SharedInputNodes.Compilation, NodeStateTable<Compilation>.WithSingleItem(compilation, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.AnalzerConfigOptions, NodeStateTable<AnalyzerConfigOptionsProvider>.WithSingleItem(_state.OptionsProvider, EntryState.Added));
+            driverStateBuilder.SetInputState(SharedInputNodes.ParseOptions, NodeStateTable<ParseOptions>.WithSingleItem(_state.ParseOptions, EntryState.Added));
+
+            // build the updated node for the syntax trees
+            PooledHashSet<SyntaxTree> syntaxTrees = PooledHashSet<SyntaxTree>.GetInstance();
+            foreach (var tree in compilation.SyntaxTrees)
             {
-                foreach (var tree in compilation.SyntaxTrees)
+                syntaxTrees.Add(tree);
+            }
+
+            var previousTrees = state.StateTable.GetStateTable<SyntaxTree>(SharedInputNodes.SyntaxTrees);
+            var newTrees = new NodeStateTable<SyntaxTree>.Builder();
+            foreach ((var tree, var treeState) in previousTrees)
+            {
+                newTrees.AddEntriesFromPreviousTable(previousTrees, syntaxTrees.Contains(tree) ? EntryState.Cached : EntryState.Removed);
+                syntaxTrees.Remove(tree);
+            }
+            foreach (var tree in syntaxTrees)
+            {
+                newTrees.AddEntries(ImmutableArray.Create(tree), EntryState.Added);
+            }
+            var syntaxTreeTable = newTrees.ToImmutableAndFree();
+            driverStateBuilder.SetInputState(SharedInputNodes.SyntaxTrees, syntaxTreeTable);
+            syntaxTrees.Free();
+
+            // PROTOTYPE(source-generators): Can we do this inline as we update the trees above?
+            if (syntaxInputNodes.Count > 0 || receiverCount > 0)
+            {
+                var builders = syntaxInputNodes.SelectAsArray(n => n.GetBuilder(_state.StateTable));
+                foreach ((var tree, var treeState) in driverStateBuilder.GetLatestStateTableForNode(SharedInputNodes.SyntaxTrees))
                 {
-                    var root = tree.GetRoot(cancellationToken);
+                    if (treeState == EntryState.Removed)
+                    {
+                        // building a new state, so don't need to worry about the removed.
+                        continue;
+                    }
+
+                    // PROTOTYPE(source-generators): It would be nice to somehow collect all the results up front, then pass them to the builder
+                    //                               so we don't have to worry about the generic and have the extra interface.
+
                     var semanticModel = compilation.GetSemanticModel(tree);
+                    Lazy<SyntaxNode> root = new Lazy<SyntaxNode>(() => tree.GetRoot(cancellationToken));
```

Comment:
>Lazy<SyntaxNode> root = new Lazy<SyntaxNode>(() => tree.GetRoot(cancellationToken)) [](http://example.com/codeflow?start=20&length=83)

Can we avoid creating the `Lazy<SyntaxNode>` instance and the delegate and closure and simply call `tree.GetRoot(cancellationToken)` in the two locations where we need it below? (The `SyntaxTree` presumably handles caching the result already.)